### PR TITLE
feat: AdGuard Home integration

### PR DESCRIPTION
**Pull request by hillaliy**
<br/>
<div align="center">
  
  <h3>Homarr</h3>
</div>

**Thank you for your contribution. Please ensure that your pull request meets the following pull request:**

- [x] Builds without warnings or errors (``pnpm buid``, autofix with ``pnpm format:fix``)
- [x] Pull request targets ``dev`` branch
- [x] Commits follow the [conventional commits guideline](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] No shorthand variable names are used (eg. ``x``, ``y``, ``i`` or any abbrevation)

### CODE REVIEW

File: packages/api/src/router/widgets/dns-hole.ts [hunk 1]
```diff
@@ -22,9 +22,9 @@ export const dnsHoleRouter = createTRPCRouter({
               case "piHole":
                 client = new PiHoleIntegration(integration);
                 break;
-              // case 'adGuardHome':
-              //   client = new AdGuardHomeIntegration(integration);
-              //   break;
+              case "adGuardHome":
+                client = new AdGuardHomeIntegration(integration);
+                break;
               default:
```
Comment (Meierschlumpf) [inline packages/api/src/router/widgets/dns-hole.ts:28]: default option can be removed as we restrict only piHole and adGuardHome in the middleware (`unstable_concat`)
Please update for all 3 occurences

File: packages/integrations/src/adguard-home/adguard-home-integration.ts [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+      },
+    });
+
+    if (!statusResponse.ok) {
+      throw new Error(
+        `Failed to fetch status for ${this.integration.name} (${this.integration.id}): ${statusResponse.statusText}`,
+      );
+    }
+
+    const filteringStatusresponse = await fetch(`${this.integration.url}/control/filtering/status`, {
```
Comment (Meierschlumpf) [inline packages/integrations/src/adguard-home/adguard-home-integration.ts:35]: change to `filteringStatusResponse`

File: packages/integrations/src/adguard-home/adguard-home-integration.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+      );
+    }
+
+    const stats = statsResponseSchema.safeParse(await statsResponse.json());
+    const status = statusResponseSchema.safeParse(await statusResponse.json());
+    const filteringStatus = filteringStatusSchema.safeParse(await filteringStatusresponse.json());
+
+    if (!stats.success) {
+      throw new Error(
+        `Failed to parse summary for ${this.integration.name} (${this.integration.id}), most likely your username or password is wrong: ${stats.error.message}`,
```
Comment (Meierschlumpf) [inline packages/integrations/src/adguard-home/adguard-home-integration.ts:53]: Add which one was not successful (here status)

File: packages/integrations/src/adguard-home/adguard-home-integration.ts [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+  public async enableAsync(): Promise<void> {
+    const username = super.getSecretValue("username");
+    const password = super.getSecretValue("password");
+    const response = await fetch(`${this.integration.url}/control/protection`, {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/json",
+        Authorization: `Basic ${Buffer.from(`${username}:${password}`).toString("base64")}`,
```
Comment (Meierschlumpf) [inline packages/integrations/src/adguard-home/adguard-home-integration.ts:122]: Maybe you could add a private method that generates the authorization header so the logic of getting username, password and converting them to base64 has not to be done 4 times

File: packages/integrations/src/adguard-home/adguard-home-integration.ts [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    duration = duration !== undefined ? duration * 1000 : 0;
+    const response = await fetch(`${this.integration.url}/control/protection`, {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/json",
+        Authorization: `Basic ${Buffer.from(`${username}:${password}`).toString("base64")}`,
+      },
+      body: JSON.stringify({
+        enabled: false,
+        duration: duration,
```
Comment (Meierschlumpf) [inline packages/integrations/src/adguard-home/adguard-home-integration.ts:147]:
```suggestion
        duration,
```

**Review by Meierschlumpf — CHANGES_REQUESTED** [review]

File: packages/integrations/src/adguard-home/adguard-home-integration.ts [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    if (!stats.success) {
+      errorMessage += `\n- Stats parsing error: ${stats.error.message}`;
+    }
+    if (!status.success) {
+      errorMessage += `\n- Status parsing error: ${status.error.message}`;
+    }
+    if (!filteringStatus.success) {
+      errorMessage += `\n- Filtering status parsing error: ${filteringStatus.error.message}`;
+    }
+    if (!stats.success || !status.success || !filteringStatus.success) {
```
Comment (Meierschlumpf) [inline packages/integrations/src/adguard-home/adguard-home-integration.ts:59]: Very nice that you combined it to one error thrown. Maybe you could use an array for the error Message and check here that the length is bigger than 2. Then you ca just call `errorMessages.join("\n")`

**Review by Meierschlumpf — CHANGES_REQUESTED** [review]
Other than that looks very nice 👍

File: packages/integrations/src/adguard-home/adguard-home-integration.ts [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    if (!stats.success) {
+      errorMessages.push(`Stats parsing error: ${stats.error.message}`);
+    }
+    if (!status.success) {
+      errorMessages.push(`Status parsing error: ${status.error.message}`);
+    }
+    if (!filteringStatus.success) {
+      errorMessages.push(`Filtering status parsing error: ${filteringStatus.error.message}`);
+    }
+    if (!stats.success || !status.success || !filteringStatus.success) {
```
Comment (Meierschlumpf) [inline packages/integrations/src/adguard-home/adguard-home-integration.ts:59]: Instead you could check the length of errorMessages like `errorMessages.length >= 1`

**Review by Meierschlumpf — CHANGES_REQUESTED** [review]

File: packages/integrations/src/adguard-home/adguard-home-integration.ts (hunk 7 above)
Comment (hillaliy) [reply packages/integrations/src/adguard-home/adguard-home-integration.ts:59]: It's causes typescript error: 'stats.data' is possibly 'undefined' .....

Comment (Meierschlumpf) [reply packages/integrations/src/adguard-home/adguard-home-integration.ts:59]: Oh I see, makes sense then 👍

**Review by Meierschlumpf — APPROVED** [review]

**Pull request merged**